x/lockup/keeper/lock_test.go

```diff
@@ -7,8 +7,105 @@ import (
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )
 
+func (suite *KeeperTestSuite) TestRebondTokens() {
+	// coins to lock
+	coins := sdk.NewCoins(sdk.NewInt64Coin("stake", 10))
+	// coins to rebond
+	coinsRebond := sdk.NewCoins(sdk.NewInt64Coin("stake", 5))
+
+	addr1 := suite.TestAccs[0]
+	defaultLockID := uint64(1)
+	nonExistingLockID := uint64(2)
+
+	testCases := []struct {
+		name          string
+		unlock        bool
+		rebondLockID  uint64
+		coinsToRebond sdk.Coins
+
+		expectedError error
+	}{
+		{
+			name:          "Valid: rebond some tokens",
+			unlock:        true,
+			rebondLockID:  defaultLockID,
+			coinsToRebond: coinsRebond,
+		},
+		{
+			name:          "Valid: explicitly rebond all tokens",
+			unlock:        true,
+			rebondLockID:  defaultLockID,
+			coinsToRebond: coins,
+		},
+		{
+			name:         "Valid: implicitly rebond all tokens by setting coinsToRebond as nil",
+			unlock:       true,
+			rebondLockID: defaultLockID,
+		},
+		{
+			name:          "Invalid: Trying to rebond a non existent lock id",
+			unlock:        true,
+			rebondLockID:  nonExistingLockID,
+			coinsToRebond: coinsRebond,
+			expectedError: sdkerrors.Wrap(types.ErrLockupNotFound, fmt.Sprintf("lock with ID %d does not exist", nonExistingLockID)),
+		},
+		{
+			name:          "Invalid: Trying to rebond a non-unbonding lock",
+			unlock:        false,
+			rebondLockID:  defaultLockID,
+			coinsToRebond: coinsRebond,
+			expectedError: fmt.Errorf("lock %d is not unlocking, rebonding only possible in unlocking stage", defaultLockID),
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			// lock coins
+			lockID := suite.LockTokens(addr1, coins, time.Second)
+
+			if tc.unlock {
+				// unlock coins
+				_, err := suite.App.LockupKeeper.BeginUnlock(suite.Ctx, lockID, coins)
+				suite.Require().NoError(err)
+			}
+
+			// rebond coins
+			err := suite.App.LockupKeeper.RebondTokens(suite.Ctx, tc.rebondLockID, addr1, tc.coinsToRebond)
+
+			if tc.expectedError != nil {
+				suite.Require().EqualError(err, tc.expectedError.Error())
+				return
+			}
+			suite.Require().NoError(err)
+
+			// get original lock
+			lock, err := suite.App.LockupKeeper.GetLockByID(suite.Ctx, lockID)
+			suite.Require().NoError(err)
+
+			if tc.coinsToRebond != nil && !coins.IsEqual(tc.coinsToRebond) {
+				// check original lock: should have less coins and be in unlocking state
+				suite.Require().Equal(lock.Coins, coins.Sub(tc.coinsToRebond))
+				suite.Require().True(lock.IsUnlocking())
+
+				// check newly created lock
+				lastLockID := suite.App.LockupKeeper.GetLastLockID(suite.Ctx)
+				rebondedLock, err := suite.App.LockupKeeper.GetLockByID(suite.Ctx, lastLockID)
+				suite.Require().NoError(err)
+				suite.Require().Equal(rebondedLock.Coins, tc.coinsToRebond)
+				suite.Require().False(rebondedLock.IsUnlocking())
+			} else {
```

Comment:
also can we add check that other fields for the locks remain changed / unchanged?